Relax AVX512 requirements for older GCC support

dev0x13 opened the pull request:
'_mm512_set_epi8' intrinsic is missing in GCC <9. However, its usage can be easily eliminated in favor of '_mm512_set_epi32' with no loss in performance enabling older GCC to benefit from AVX512-optimized codepaths.

Dead2 commented on configure:
```diff
@@ -1528,6 +1543,14 @@ case "${ARCH}" in
+            if test ${MTUNE_CASCADELAKE_AVAILABLE} -eq 1; then
+                mtuneflag="-mtune=cascadelake"
+            else
+                mtuneflag="-mtune=skylake-avx512"
```
I am pretty sure this would fail badly on RHEL7 GCC 4.8 for example.

This should probably be within the `HAVE_AVX512_INTRIN` if block above, otherwise even an i386 cpu would get this mtune set.

I also think calling this `mtuneflag` is misleading, since it only applies to AVX512 files.

How about calling it `avx512tune` and letting it default to `skylake-avx512` at line 111, then only change it if the above check is successful.

I am assuming that you made sure that `-mtune=skylake-avx512` works for all GCC/Clang compilers supporting AVX512 in any capacity..

Dead2 replied:
Sorry, didn't notice this was listed as 'pending' and thus not visible to anyone but me..

mtl1979 replied:
We don't support i386 as target... Basically anything older than Pentium Pro (i686) is unsupported, as we don't/can't do checks if MMX is available or not (cpuid instruction is not even available)... For most modern 64-bit processors MMX is irrelevant as SSE and AVX superceded it, but for real 32-bit processors it makes a big difference. I do agree that the choice for variable name isn't best possible, even though only one "-mtune" flag should ever be used on any single file... However we should test that both options are supported before explicitly enabling either one of them.

dev0x13 commented:
@Dead2 May I ask which changes are requested?

Dead2 commented:
> @Dead2 May I ask which changes are requested?

That would be the ones I requested in my review, that can be seen about half a page above this comment.

nmoinvaz commented:
I'm not seeing the review.

nmoinvaz commented:
The build changes and code changes should be split into separate commits. I'm not sure the build changes are needed - does the commit mention why they are necessary?

dev0x13 commented:
I don't see a review either.
I'll split the changes. Build changes are required too because the AVX512 detection is based on probing simple C programs containing the corresponding intrinsics.

KungFuJesus commented:
While I'm sure this code is functionally correct - I do not love that it impedes legibility.  Is there any possible way to write a macro that uses a fallback method to derive these same constants at compile time?  That just makes it much more obvious what's going on and if someone wants to go back and change the implementation to say, use a different mixing matrix of sorts, they don't have to struggle using the set_epi32 intrinsic to derive the proper constants.

KungFuJesus commented on arch/x86/adler32_avx512_tpl.h:
```diff
@@ -46,10 +46,10 @@ Z_INTERNAL uint32_t adler32_avx512(uint32_t adler, const uint8_t *src, uint64_t
-                                          20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37,
-                                          38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
-                                          56, 57, 58, 59, 60, 61, 62, 63, 64);
+    const __m512i dot2v = _mm512_set_epi32(0x1020304, 0x5060708, 0x90a0b0c, 0xd0e0f10,
```
While I'm sure this code is functionally correct - I do not love that it impedes legibility. Is there any possible way to write a macro that uses a fallback method to derive these same constants at compile time? That just makes it much more obvious what's going on and if someone wants to go back and change the implementation to say, use a different mixing matrix of sorts, they don't have to struggle using the set_epi32 intrinsic to derive the proper constants.

KungFuJesus replied:
There's also the fact that GCC prior to 9 kind of emitted garbage code for some AVX512 intrinsics, but that's another matter.  It was missing a lot of keyhole optimizations with regard to masks (and it's still not great, it zealously moves masks into GPRs a bit too often).

nmoinvaz replied:
What do you think about doing `#ifdef __GNUC__ < 9` with both versions in the source?

KungFuJesus replied:
Doable, but it might be cleaner and more future proof to just have selectively bit shifts per byte (and 16 bitwise ORs) on gcc versions where the intrinsic isn't available.  Getting that right means we should be able to specify exact uint8 values per lane should the implementation change or AVX512 gets used elsewhere.

mtl1979 replied:
We have header(s) for "fallback" functions for a reason... If compiler doesn't support something, we should add any replacement function to most relevant fallback header. 

All fallback functions should be generic with no expectations on what parameters are passed to them. Any repeated calculations should be added as macros to make the code as readable as possible. As there is only little-endian x86 processors, there only needs to be one variation of the macros. Care has to be taken with preprocessor macros as normal integer promotion and conversion rules do not apply, so shifting must mask any extra bits.

dev0x13 committed:
Add fallback function for '_mm512_set_epi8' intrinsic

'_mm512_set_epi8' intrinsic is missing in GCC <9.
However, its usage can be easily eliminated in
favor of '_mm512_set_epi32' with no loss in
performance enabling older GCC to benefit from
AVX512-optimized codepaths.

dev0x13 committed:
Enable using AVX512 intrinsics with GCC <9

Replace missing '_mm512_set_epi8' with
'_mm512_set_epi32' in test code for configuring;
Add fallback for '-mtune=cascadelake' flag used
when AVX512 is enabled.

dev0x13 commented:
@Dead2 @nmoinvaz @KungFuJesus @mtl1979 
Hi gentlemen. Apologies for disappearing earlier - couldn't make it to non-work coding.
I believe I've made all the changes requested, namely:
1. Separated changes into two commits: one for code and one for build configuration.
2. Replaced obscure `_mm512_set_epi32` invocation with an easy-to-use fallback function.
3. Change the misleading `mtuneflag` name in the Makefile to something more appropriate.

KungFuJesus reviewed (approved):
LGTM.  The cascade lake mtune did show some measurable improvement in instruction scheduling for this code from what I recall but you should still end up with something faster than not having the AVX512 variants.

VNNI is actually quite a bit better but if you're dealing with GCCs this old you won't have that.

Dead2 reviewed (approved):
LGTM

Dead2 commented on configure:
```diff
@@ -1567,6 +1597,16 @@ case "${ARCH}" in
+                avx512tune="-mtune=cascadelake"
+            else
+                if test ${MTUNE_SKYLAKE_AVX512_AVAILABLE} -eq 1; then
+                    avx512tune="-mtune=skylake-avx512"
```
After having another look at this, I am wondering; Won't `AVX512TUNE` be missing from the Makefiles and thus never actually get used for the configure flow?

I think the best solution here is not to add `AVX512TUNE` but to rather append the chosen mtune parameter to the existing `avx512flag` variable. Those two would always be used together anyways so there is probably no real need to separate them.
That also matches better with how it is done in cmake and how configure applied the mtune before, making it a minimal change.


While you are at it, I'd appreciate it if you would rebase the whole PR on top of latest develop so the last CI tests hopefully gets fixed.

dev0x13 replied:
Thanks for review! Both things are done.

Dead2 reviewed (changes requested):
Requested changes to configure
